astropy/units/quantity.py

```diff
@@ -1349,26 +1349,26 @@ def _to_own_unit(self, value, check_precision=True):
             # We're not a Quantity, so let's try a more general conversion.
             # Plain arrays will be converted to dimensionless in the process,
             # but anything with a unit attribute will use that.
+            as_quantity = Quantity(value)
             try:
-                _value = Quantity(value).to_value(self.unit)
-            except UnitsError as exc:
+                _value = as_quantity.to_value(self.unit)
+            except UnitsError:
                 # last chance: if this was not something with a unit
                 # and is all 0, inf, or nan, we treat it as arbitrary unit.
                 if (not hasattr(value, 'unit') and
-                        can_have_arbitrary_unit(value)):
-                    _value = value
+                        can_have_arbitrary_unit(as_quantity.value)):
+                    _value = as_quantity.value
                 else:
-                    raise exc
+                    raise
 
         if check_precision:
```

Comment:
Note that I am not sure this whole check is really needed, at least not for setting items (numpy doesn't do it). But that's probably best left for a separate discussion.